### PR TITLE
Secure network private key file and data directory.

### PR DESCRIPTION
**Pull request by cheatfate**
This is first way of fixes which partially fixes 80% of #1320
Right now the only missing thing is 
> consider encrypting cached key material with a user unlock/or application specific key. Scenario: other process running as the user accesses privkey (generally hard to prevent this scenario and may arguably always be bypassed unless some level of interaction is required)

### CODE REVIEW

**Comment by tersec** [comment]
#1384 is "convert 22 v0.12.1 spec refs to v0.12.2"

**Comment by cheatfate** [comment]
@tersec you are right, it was mistake proper issue is #1320, sorry.

**Comment by zah** [comment]
We can easily use a keystore for the persistent network keys as well (used for bootstrap nodes). This would still require a password file stored somewhere in order to handle the potential supervisor restarts non-interactively. Otherwise, In the recent meetings, we've discussed that most of the clients can generate and use ephemeral keys.

**Comment by cheatfate** [comment]
@zah this is my next step, i'm going to refactor `spec/keystore` to fix non constant-time hexadecimal conversions and add `createNetKeystore` procedure too.

**Comment by cheatfate** [comment]
https://github.com/status-im/nim-stew/pull/53 is prerequisite for this PR.

File: Makefile [hunk 1]
```diff
@@ -188,7 +188,7 @@ testnet0 testnet1: | beacon_node signing_process
 		$(GOERLI_TESTNETS_PARAMS) $(NODE_PARAMS)
 
 define connect_to_network
-	mkdir -p build/data/shared_$(1)_$(NODE_ID)
+	mkdir -m 0750 -p build/data/shared_$(1)_$(NODE_ID)
```
Comment (zah) [inline Makefile:191]: What do these do on Windows btw?

File: Makefile [hunk 2]
```diff
@@ -310,8 +310,7 @@ clean-medalla:
 spadina: | beacon_node signing_process
 	$(call connect_to_network,spadina)
 
-spadina-vc: | beacon_node signing_process validator_client
-	$(call connect_to_network_with_validator_client,spadina)
```
Comment (zah) [inline Makefile:314]: Was this removed by accident?

File: beacon_chain/keystore_management.nim [hunk 3]
```diff
@@ -22,6 +25,97 @@ type
...
+      true
+  else:
+    fatal "Unsupported operation system"
+    return false
+
+proc checkFilePermissions*(filePath: string): bool =
+  ## Check if ``filePath`` has only "(600) rw-------" permissions.
+  ## Procedure returns ``false`` if permissions are different
+  when defined(windows):
+    # Windows do not support per-user/group/other permissions,
```
Comment (zah) [inline beacon_chain/keystore_management.nim:93]: Hrm, Windows still has access lists that can make the file readable only by the right user.

File: beacon_chain/keystore_management.nim [hunk 4]
```diff
@@ -22,6 +25,97 @@ type
...
+        false
+      else:
+        true
+    else:
+      true
+  else:
+    fatal "Unsupported operation system"
+    return false
+
+proc checkFilePermissions*(filePath: string): bool =
```
Comment (zah) [inline beacon_chain/keystore_management.nim:89]: This name is a bit vague. It could have been `checkSensitiveFilePermissions` or even better `isFileWorldReadable` or `isAccessRestrictedToCurrentUser`.

File: beacon_chain/eth2_network.nim [hunk 5]
```diff
@@ -266,6 +266,9 @@ const
 when libp2p_pki_schemes != "secp256k1":
   {.fatal: "Incorrect building process, please use -d:\"libp2p_pki_schemes=secp256k1\"".}
 
+const
+  NetworkInsecureKeyPassword = "INSECUREPASSWORD"
```
Comment (zah) [inline beacon_chain/eth2_network.nim:270]: Can you clarify the purpose of this? (and the matching `netKeyInsecurePassword` option)

File: beacon_chain/conf.nim [hunk 6]
```diff
@@ -84,6 +85,18 @@ type
       desc: "Do not display interative prompts. Quit on missing configuration"
       name: "non-interactive" }: bool
 
+    netKeyFile* {.
+      defaultValue: "random",
+      desc: "Source of network (secp256k1) private key file " &
+            "(random|<path>) (default: random)"
+      name: "netkey-file" }: string
```
Comment (zah) [inline beacon_chain/conf.nim:92]: You can use `Option[InputFile]` here instead of `string`. Confutils will then verify that the supplied path is valid when `Config.load` is executed. The other benefit will be that Confutils has a context-aware auto-completion algorithm that can suggest a file path in the shell while the user is composing a command.

File: beacon_chain/keystore_management.nim [hunk 7]
```diff
@@ -117,6 +216,112 @@ type
...
+  if insecurePwd.isSome():
+    warn "Using insecure password to unlock networking key"
+    let decrypted = decryptNetKeystore(keystore, KeystorePass insecurePwd.get())
+    if decrypted.isOk:
+      return some(decrypted.get())
+    else:
+      error "Network keystore decryption failed", key_store = keyStorePath
+      return
+  else:
+    var remainingAttempts = 3
```
Comment (zah) [inline beacon_chain/keystore_management.nim:247]: The logic here could be aware of the `--non-interactive` command line option that should trigger an error if the code reaches here. The purpose of the `--non-interactive` flag is to avoid the situation where due to mis-configuration, the process hangs forever waiting for some user input that will never arrive.

File: beacon_chain/eth2_network.nim [hunk 8]
```diff
@@ -1191,21 +1194,117 @@ proc initAddress*(T: type MultiAddress, str: string): T =
...
+      let keyPath =
+        if isAbsolute(conf.netKeyFile):
+          conf.netKeyFile
+        else:
+          conf.dataDir / conf.netKeyFile
+
+      if fileAccessible(keyPath, {AccessFlags.Find}):
+        info "Network key storage is present, unlocking", key_path = keyPath
+
+        # Insecure password used only for automated testing.
```
Comment (zah) [inline beacon_chain/eth2_network.nim:1221]: We need to support running a bootstrap node server that can be restarted in non-interactive way. Such a setup would use a persistent network keyfile and it would be nice if it can be password protected with a password stored in the specified `secretsDir`.

File: beacon_chain/eth2_network.nim [hunk 9]
```diff
@@ -1214,23 +1313,26 @@ func gossipId(data: openArray[byte]): string =
...
+                     conf: BeaconNodeConf,
+                     netKeys: KeyPair,
+                     enrForkId: ENRForkID): Eth2Node =
   var
     (extIp, extTcpPort, extUdpPort) = setupNat(conf)
     hostAddress = tcpEndPoint(conf.libp2pAddress, conf.tcpPort)
     announcedAddresses = if extIp.isNone(): @[]
                          else: @[tcpEndPoint(extIp.get(), extTcpPort)]
 
+  let networkPublicKey = byteutils.toHex(netKeys.pubkey.getBytes().tryGet())
```
Comment (zah) [inline beacon_chain/eth2_network.nim:1326]: If the code was written like this:

```nim
  info "Initializing networking",
       hostAddress,
       networkPublicKey = byteutils.toHex(netKeys.pubkey.getBytes().tryGet()),
       announcedAddresses
```

The `networkPublicKey` would not be computed if the log level settings filter the whole statement.

You can also use the `formatIt` feature of Chronicles to automatically insert the `toHex` call every time a pubkey is logged.

File: beacon_chain/keystore_management.nim [hunk 10]
```diff
@@ -117,6 +216,112 @@ type
...
+    warn "Using insecure password to lock networking key",
+         key_path = keyStorePath
+    password = insecurePwd.get()
+  else:
+    while true:
+      let prompt = "Please enter NEW password to lock network key storage: "
+
+      password =
+        try:
+          readPasswordFromStdin(prompt)
```
Comment (zah) [inline beacon_chain/keystore_management.nim:281]: Ditto for `--non-interactive`.

File: beacon_chain/keystore_management.nim [hunk 11]
```diff
@@ -117,6 +216,112 @@ type
...
+        except IOError:
+          error "Could not read password from stdin"
+          return err(FailedToCreateKeystoreFile)
+
+      if len(password) < minPasswordLen:
```
Comment (zah) [inline beacon_chain/keystore_management.nim:286]: We can extract this logic in a sub-routine if it's going to be shared between the network keystores and the regular validator keystores.

File: beacon_chain/spec/keystore.nim [hunk 12]
```diff
@@ -554,6 +598,29 @@ proc createCryptoField(kdfKind: KdfKind,
...
+    pubKey = privKey.getKey().get()
+    uuid = uuidGenerate().expect("Random bytes should be available")
+
+  NetKeystore(
+    crypto: cryptoField,
+    pubkey: pubKey,
+    path: path,
+    description: newClone(description),
+    uuid: $uuid,
+    version: 4
```
Comment (zah) [inline beacon_chain/spec/keystore.nim:621]: Maybe we don't need to imitate the Validator keystores here. The most minimal version would use just a `crypto` field and a `version` field. 

Setting the `version` to 4 seems a bit arbitrary and the rationale seems to be that we want to make the network key store "look like" a validator key store. This is potentially confusing for the user and may lead her to making a mistake by copying the contents from one keystore to the other.

File: beacon_chain/conf.nim (hunk 6 above)
Comment (cheatfate) [reply beacon_chain/conf.nim:92]: @zah sorry but Option[InputFile] did not follow this option's logic:
1. `random` value. Means that there no such file in filesystem and no files will be created in filesystem. Random private key will be generated and used as network key.
2. `filename` which is not present in filesystem. In this case we going to create new network keystore and store it in `filename`.
3. `filename` which is present in filesystem. In such case we going to load, decode and use keystore `filename`.

File: beacon_chain/keystore_management.nim (hunk 7 above)
Comment (cheatfate) [reply beacon_chain/keystore_management.nim:247]: @zah whole password entering procedure is already outdated and it should be refactored because there like 5 security issues already found. So all this refactoring could be done in other PR.

File: beacon_chain/conf.nim (hunk 6 above)
Comment (zah) [reply beacon_chain/conf.nim:92]: The difference is extremely subtle. Only when the user explicitly types `random`, this won't be considered a filename in the current setup, but it will be in the suggested one. Typing out `random` seems a bit unnecessary for the user when it's the default behaviour and omitting the option can quite naturally be interpreted as "I don't need a persistent filename". On the other hand, having a special "keyword" that can be mistyped seems less elegant and muddles the type system which prevents the other benefits that I mentioned. 

To recap:

1) netkeyFile.isNone is the the same as random
2) netkeyFile.isSome can point to an existing or non-existing file (that will be created)

File: beacon_chain/keystore_management.nim (hunk 7 above)
Comment (zah) [reply beacon_chain/keystore_management.nim:247]: The planned improvements are even stronger argument to extract the shared procedure now, so there is a single place in the code where the improvements can be introduced.

File: beacon_chain/conf.nim (hunk 6 above)
Comment (cheatfate) [reply beacon_chain/conf.nim:92]: that will be created by confutils?

File: beacon_chain/keystore_management.nim (hunk 7 above)
Comment (cheatfate) [reply beacon_chain/keystore_management.nim:247]: @zah please lets concentrate on number of issues which solves this PR (permissions), its already pretty big, and i really do not want to add fixes for password (which i have already made).

File: beacon_chain/eth2_network.nim (hunk 5 above)
Comment (cheatfate) [reply beacon_chain/eth2_network.nim:276]: Because keystore could not have empty password `NetworkInsecureKeyPassword` will be used instead and if you use `--netKeyInsecurePassword` option this password will be used to lock keystore.
